File: code/c/benchmark.c
/**
 * @file benchmark.c
 * @brief Matrix multiplication benchmarking harness for C implementation
 * 
 * This program benchmarks the performance of matrix multiplication across
 * multiple matrix sizes and runs, recording execution time, CPU usage, and
 * memory consumption to a CSV file.
 * 
 * Command-line arguments:
 *   argv[1]: Comma-separated matrix sizes (e.g., "64,128,256")
 *   argv[2]: Number of runs per size (default: 3)
 *   argv[3]: Output CSV file path (default: "results_raw.csv")
 *   argv[4]: Random seed (default: 27)
 * 
 * Example: benchmark.exe "64,128,256" 5 output.csv 42
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <windows.h>
#include <psapi.h>
#include "matrix_mult.h"

/* CSV header format */
#define HEADER "run_id;language;size;run_idx;time_ms;cpu_pct;peak_mib\n"

/**
 * @brief Get current time in seconds with high precision
 * @return Current time in seconds as a double
 */
static double now_sec(void) {
    LARGE_INTEGER frequency, counter;
    QueryPerformanceFrequency(&frequency);
    QueryPerformanceCounter(&counter);
    return (double)counter.QuadPart / (double)frequency.QuadPart;
}

/**
 * @brief Get total CPU time consumed by the current process
 * @return CPU time in seconds (user + kernel time)
 */
static double proc_cpu_seconds(void) {
    FILETIME creation_time, exit_time, kernel_time, user_time;
    GetProcessTimes(GetCurrentProcess(), &creation_time, &exit_time, 
                    &kernel_time, &user_time);
    
    /* Convert FILETIME to 64-bit integer (100-nanosecond intervals) */
    ULONGLONG kernel_ticks = ((ULONGLONG)kernel_time.dwLowDateTime) | 
                             ((ULONGLONG)kernel_time.dwHighDateTime << 32);
    ULONGLONG user_ticks = ((ULONGLONG)user_time.dwLowDateTime) | 
                           ((ULONGLONG)user_time.dwHighDateTime << 32);
    
    /* Convert to seconds (1e7 = 100ns per second) */
    return (kernel_ticks + user_ticks) / 1e7;
}

/**
 * @brief Get number of logical CPU cores
 * @return Number of logical processors, or 1 if detection fails
 */
static int logical_cpus(void) {
    SYSTEM_INFO system_info;
    GetSystemInfo(&system_info);
    return system_info.dwNumberOfProcessors > 0 ? 
           (int)system_info.dwNumberOfProcessors : 1;
}

/**
 * @brief Get current process memory usage in MiB
 * @return Working set size in mebibytes (MiB), or 0.0 on error
 */
static double current_mem_mib(void) {
    PROCESS_MEMORY_COUNTERS pmc;
    if (GetProcessMemoryInfo(GetCurrentProcess(), &pmc, sizeof(pmc))) {
        return pmc.WorkingSetSize / (1024.0 * 1024.0);
    }
    return 0.0;
}

/**
 * @brief Write CSV header to file if it doesn't exist
 * @param path Path to the output CSV file
 */
static void write_header_if_needed(const char* path) {
    FILE* f = fopen(path, "r");
    if (f) {
        fclose(f);
        return; /* File exists, header already written */
    }
    
    /* File doesn't exist, create it and write header */
    f = fopen(path, "a");
    fputs(HEADER, f);
    fclose(f);
}

/**
 * @brief Generate a run identifier string from current local time
 * @param buf Buffer to store the run ID string
 * @param n Size of the buffer
 * 
 * Format: DD/MM/HH/MM (day/month/hour/minute)
 */
static void run_id_str(char* buf, size_t n) {
    SYSTEMTIME st;
    GetLocalTime(&st);
    _snprintf_s(buf, n, _TRUNCATE, "%02d/%02d/%02d/%02d", 
                st.wDay, st.wMonth, st.wHour, st.wMinute);
}

/**
 * @brief Main benchmarking entry point
 * 
 * Parses command-line arguments, generates random matrices, and runs
 * the matrix multiplication benchmark for each specified size and run count.
 */
int main(int argc, char** argv) {
    /* Default configuration */
    int sizes_default[] = {64, 128, 256, 512, 1024};
    int sizes[64];  /* Maximum 64 different sizes */
    int nsizes = 0;
    int runs = 3;
    const char* out = "results_raw.csv";
    int seed = 27;
    
    /* Parse command-line argument 1: comma-separated matrix sizes */
    if (argc > 1 && argv[1][0]) {
        const char* p = argv[1];
        while (*p && nsizes < 64) {
            sizes[nsizes++] = (int)strtol(p, NULL, 10);
            p = strchr(p, ',');
            if (!p) break;
            ++p;
        }
    }
    
    /* Use default sizes if none specified */
    if (nsizes == 0) {
        memcpy(sizes, sizes_default, sizeof(sizes_default));
        nsizes = (int)(sizeof(sizes_default) / sizeof(sizes_default[0]));
    }
    
    /* Parse command-line argument 2: number of runs */
    if (argc > 2) runs = atoi(argv[2]);
    
    /* Parse command-line argument 3: output file path */
    if (argc > 3) out = argv[3];
    
    /* Parse command-line argument 4: random seed */
    if (argc > 4) seed = atoi(argv[4]);
    
    /* Initialize random number generator */
    srand(seed);
    
    /* Prepare output file */
    write_header_if_needed(out);
    
    /* Generate run identifier */
    char run_id[32];
    run_id_str(run_id, sizeof(run_id));
    const char* language = "C";
    const int ncpu = logical_cpus();
    
    /* Main benchmarking loop: iterate over all matrix sizes */
    for (int si = 0; si < nsizes; ++si) {
        int n = sizes[si];
        
        /* Allocate matrices A, B, and C */
        float* A = (float*)malloc((size_t)n * (size_t)n * sizeof(float));
        float* B = (float*)malloc((size_t)n * (size_t)n * sizeof(float));
        float* C = (float*)malloc((size_t)n * (size_t)n * sizeof(float));
        
        /* Initialize matrices with random values in [0, 1] */
        for (int i = 0; i < n * n; i++) {
            A[i] = (float)rand() / RAND_MAX;
            B[i] = (float)rand() / RAND_MAX;
        }
        
        /* Perform multiple runs for statistical stability */
        for (int r = 1; r <= runs; ++r) {
            /* Capture metrics before execution */
            double mem_before = current_mem_mib();
            double cpu0 = proc_cpu_seconds();
            double t0 = now_sec();
            
            /* Execute matrix multiplication */
            matrix_multiplication(A, B, C, n);
            
            /* Capture metrics after execution */
            double t1 = now_sec();
            double cpu1 = proc_cpu_seconds();
            double mem_after = current_mem_mib();
            
            /* Calculate performance metrics */
            double wall = t1 - t0;                          /* Wall-clock time */
            double time_ms = wall * 1000.0;                 /* Convert to milliseconds */
            double cpu_pct = 100.0 * (cpu1 - cpu0) / (wall * ncpu);  /* CPU percentage */
            double peak_mib = mem_after > mem_before ? mem_after : mem_before;
            
            /* Print results to console */
            printf("n=%d run=%d time=%.2f ms CPU=%.1f%% MEM=%.2f MiB\n", 
                   n, r, time_ms, cpu_pct, peak_mib);
            
            /* Append results to CSV file */
            FILE* f = fopen(out, "a");
            fprintf(f, "%s;%s;%d;%d;%.3f;%.1f;%.2f\n", 
                    run_id, language, n, r, time_ms, cpu_pct, peak_mib);
            fclose(f);
        }
        
        /* Free allocated matrices */
        free(A);
        free(B);
        free(C);
    }
    
    return 0;
}

File: code/python/Benchmark.py
"""
Matrix multiplication benchmarking harness for Python implementation.

This module benchmarks the performance of matrix multiplication across
multiple matrix sizes and runs, recording execution time, CPU usage, and
memory consumption to a CSV file.

Command-line arguments:
    --sizes: Space-separated matrix sizes (default: 64 128 256 512 1024)
    --runs: Number of runs per size (default: 3)
    --out: Output CSV file path (default: "results_raw.csv")
    --seed: Random seed (default: 27)
    --check_n: Matrix size for correctness check (default: 5)

Example:
    python benchmark.py --sizes 64 128 256 --runs 5 --out output.csv --seed 42
"""

import argparse
import os
import time
from typing import Tuple

import numpy as np
import psutil

from matrix_mult import matrixMultiplication

# CSV header format
HEADER = "run_id;language;size;run_idx;time_ms;cpu_pct;peak_mib\n"


def check_correctness(n: int, seed: int = 27, atol: float = 1e-8) -> bool:
    """
    Verify correctness of matrix multiplication implementation.
    
    Tests the multiplication with an n×n matrix and compares the result
    against NumPy's built-in matrix multiplication.
    
    Args:
        n: Dimension of the square matrix
        seed: Random seed for reproducibility
        atol: Absolute tolerance for element-wise comparison
    
    Returns:
        True if implementation is correct, False otherwise
    """
    rng = np.random.default_rng(seed)
    A = rng.random((n, n), dtype=np.float64)
    B = rng.random((n, n), dtype=np.float64)
    
    # Compare custom implementation against NumPy's optimized version
    return np.allclose(A @ B, matrixMultiplication(A, B), atol=atol)


def one_run(
    A: np.ndarray, 
    B: np.ndarray, 
    proc: psutil.Process, 
    ncpu: int
) -> Tuple[float, float, float]:
    """
    Execute a single matrix multiplication run and collect metrics.
    
    Args:
        A: First input matrix (n×n)
        B: Second input matrix (n×n)
        proc: psutil Process object for the current process
        ncpu: Number of logical CPU cores
    
    Returns:
        Tuple of (execution_time_ms, cpu_percentage, peak_memory_mib)
    """
    # Capture metrics before execution
    mem_before = proc.memory_info().rss / (1024 * 1024)  # Convert bytes to MiB
    cpu_before = proc.cpu_times()
    t0 = time.perf_counter()
    
    # Execute matrix multiplication
    matrixMultiplication(A, B)
    
    # Capture metrics after execution
    t1 = time.perf_counter()
    cpu_after = proc.cpu_times()
    mem_after = proc.memory_info().rss / (1024 * 1024)  # Convert bytes to MiB
    
    # Calculate performance metrics
    wall = max(t1 - t0, 1e-12)  # Avoid division by zero
    cpu_used = (cpu_after.user - cpu_before.user) + (cpu_after.system - cpu_before.system)
    cpu_pct = 100.0 * cpu_used / (wall * ncpu)
    
    return wall * 1000.0, cpu_pct, max(mem_before, mem_after)


def write_header_if_needed(path: str) -> None:
    """
    Write CSV header to file if it doesn't already exist.
    
    Args:
        path: Path to the output CSV file
    """
    if not os.path.isfile(path):
        with open(path, "w", encoding="utf-8") as f:
            f.write(HEADER)


def main() -> None:
    """
    Main benchmarking entry point.
    
    Parses command-line arguments, verifies correctness, generates random
    matrices, and runs the matrix multiplication benchmark for each
    specified size and run count.
    """
    # Parse command-line arguments
    p = argparse.ArgumentParser(description="Raw-run Python benchmark.")
    p.add_argument("--sizes", type=int, nargs="+", default=[64, 128, 256, 512, 1024],
                   help="Matrix sizes to benchmark")
    p.add_argument("--runs", type=int, default=3,
                   help="Number of runs per matrix size")
    p.add_argument("--out", type=str, default="results_raw.csv",
                   help="Output CSV file path")
    p.add_argument("--seed", type=int, default=27,
                   help="Random seed for reproducibility")
    p.add_argument("--check_n", type=int, default=5,
                   help="Matrix size for correctness verification")
    args = p.parse_args()

    # Verify implementation correctness before benchmarking
    if not check_correctness(args.check_n, args.seed):
        raise SystemExit("Verification failed")

    # Initialize components
    rng = np.random.default_rng(args.seed)
    proc = psutil.Process(os.getpid())
    ncpu = psutil.cpu_count(logical=True) or 1
    language = "Python"
    run_id = time.strftime("%d/%m/%H/%M", time.localtime())

    # Prepare output file
    write_header_if_needed(args.out)

    # Main benchmarking loop: iterate over all matrix sizes
    for n in args.sizes:
        # Generate random matrices for this size
        A = rng.random((n, n), dtype=np.float32)
        B = rng.random((n, n), dtype=np.float32)
        
        # Perform multiple runs for statistical stability
        for r in range(1, args.runs + 1):
            # Execute run and collect metrics
            t_ms, cpu_pct, peak_mib = one_run(A, B, proc, ncpu)
            
            # Print results to console
            print(f"n={n} run={r} time={t_ms:.2f} ms CPU={cpu_pct:.1f}% MEM={peak_mib:.2f} MiB")
            
            # Append results to CSV file
            with open(args.out, "a", encoding="utf-8") as f:
                f.write(f"{run_id};{language};{n};{r};{t_ms:.3f};{cpu_pct:.1f};{peak_mib:.2f}\n")


if __name__ == "__main__":
    main()